### PR TITLE
Gradient clipping

### PR DESCRIPTION
**Pull request by jeffdonahue**
This lets you add a field `clip_gradients` to a `SolverParameter` to set the maximum L2 norm of the parameter gradients.  If the norm of the gradients exceeds the value set, the gradients are scaled down to have exactly that L2 norm.  The L2 norm is computed prior to applying learning rate, etc. -- it applies to the actual loss gradient.

### CODE REVIEW

File: src/caffe/blob.cpp [hunk 1]
```diff
@@ -205,6 +205,142 @@ Dtype Blob<Dtype>::asum_diff() const {
...
+  return 0;
+}
+
+template <> int Blob<int>::sumsq_data() const {
+  NOT_IMPLEMENTED;
+  return 0;
+}
+
+template <typename Dtype>
+Dtype Blob<Dtype>::sumsq_data() const {
```
Comment (jeffdonahue) [inline src/caffe/blob.cpp]: I don't like writing or reading these duplicate methods for `data` and `diff` -- should these be in `SyncedMemory`, or perhaps a new abstraction layer between `SyncedMemory` and `Blob`?

**Comment by jeffdonahue** [comment]
Merging after discussion with @shelhamer.
